File: src/ru/hudspeed/pro/SpeedControlRules.java
package ru.gpsantiradar.app;

import java.util.Calendar;
import java.util.Locale;

/** Encodes conditional RadarBase speed controls without adding a second SQL table. */
public final class SpeedControlRules {
    public static final int CAR = 1;
    public static final int BIKE = 2;
    public static final int TRUCK = 4;
    public static final int BUS = 8;

    private SpeedControlRules() {}

    public static String encode(int speed, boolean average, int fromMinute, int toMinute,
                                int weekdayMask, int monthMask, int vehicleMask) {
        return speed + "," + (average ? 1 : 0) + "," + fromMinute + "," + toMinute
                + "," + weekdayMask + "," + monthMask + "," + vehicleMask;
    }

    public static int currentCarLimit(String encoded, long timeMillis) {
        if (encoded == null || encoded.isEmpty()) return 0;
        Calendar now = Calendar.getInstance();
        now.setTimeInMillis(timeMillis);
        int result = select(encoded, now, false);
        return result > 0 ? result : select(encoded, now, true);
    }

    private static int select(String encoded, Calendar now, boolean average) {
        int result = 0;
        for (String item : encoded.split(";")) {
            String[] fields = item.split(",", -1);
            if (fields.length != 7) continue;
            try {
                int speed = Integer.parseInt(fields[0]);
                boolean ruleAverage = "1".equals(fields[1]);
                int from = Integer.parseInt(fields[2]);
                int to = Integer.parseInt(fields[3]);
                int weekdays = Integer.parseInt(fields[4]);
                int months = Integer.parseInt(fields[5]);
                int vehicles = Integer.parseInt(fields[6]);
                if (speed <= 0 || ruleAverage != average) continue;
                if (vehicles != 0 && (vehicles & CAR) == 0) continue;
                if (!matchesDate(now, from, to, weekdays, months)) continue;
                if (result == 0 || speed < result) result = speed;
            } catch (NumberFormatException ignored) {
                // Ignore one malformed rule but keep the object usable.
            }
        }
        return result;
    }

    private static boolean matchesDate(Calendar now, int from, int to,
                                       int weekdays, int months) {
        int monthBit = 1 << now.get(Calendar.MONTH);
        if (months != 0 && (months & monthBit) == 0) return false;
        int mondayFirstDay = (now.get(Calendar.DAY_OF_WEEK) + 5) % 7;
        if (weekdays != 0 && (weekdays & (1 << mondayFirstDay)) == 0) return false;
        if (from < 0 || to < 0 || from == to) return true;
        int minute = now.get(Calendar.HOUR_OF_DAY) * 60 + now.get(Calendar.MINUTE);
        return from < to ? minute >= from && minute < to : minute >= from || minute < to;
    }

    public static int vehicleMask(String vehicle) {
        if (vehicle == null) return 0;
        switch (vehicle.toLowerCase(Locale.US)) {
            case "car": return CAR;
            case "bike": return BIKE;
            case "truck": return TRUCK;
            case "bus": return BUS;
            default: return 16;
        }
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/SpeedControlRules.kt
package ru.gpsantiradar.app

import java.util.Calendar
import java.util.Locale

/** Encodes conditional RadarBase speed controls without adding a second SQL table. */
object SpeedControlRules {
    const val CAR = 1
    const val BIKE = 2
    const val TRUCK = 4
    const val BUS = 8

    fun encode(
        speed: Int, average: Boolean, fromMinute: Int, toMinute: Int,
        weekdayMask: Int, monthMask: Int, vehicleMask: Int
    ) = "$speed,${if (average) 1 else 0},$fromMinute,$toMinute,$weekdayMask,$monthMask,$vehicleMask"

    fun currentCarLimit(encoded: String?, timeMillis: Long): Int {
        if (encoded.isNullOrEmpty()) return 0
        val now = Calendar.getInstance().apply { this.timeInMillis = timeMillis }
        return select(encoded, now, false).takeIf { it > 0 } ?: select(encoded, now, true)
    }

    private fun select(encoded: String, now: Calendar, average: Boolean): Int {
        var result = 0
        for (item in encoded.split(';')) {
            val fields = item.split(',')
            if (fields.size != 7) continue
            try {
                val speed = fields[0].toInt()
                val ruleAverage = fields[1] == "1"
                val from = fields[2].toInt()
                val to = fields[3].toInt()
                val weekdays = fields[4].toInt()
                val months = fields[5].toInt()
                val vehicles = fields[6].toInt()
                if (speed <= 0 || ruleAverage != average) continue
                if (vehicles != 0 && vehicles and CAR == 0) continue
                if (!matchesDate(now, from, to, weekdays, months)) continue
                if (result == 0 || speed < result) result = speed
            } catch (_: NumberFormatException) { }
        }
        return result
    }

    private fun matchesDate(now: Calendar, from: Int, to: Int, weekdays: Int, months: Int): Boolean {
        val monthBit = 1 shl now.get(Calendar.MONTH)
        if (months != 0 && months and monthBit == 0) return false
        val mondayFirstDay = (now.get(Calendar.DAY_OF_WEEK) + 5) % 7
        if (weekdays != 0 && weekdays and (1 shl mondayFirstDay) == 0) return false
        if (from < 0 || to < 0 || from == to) return true
        val minute = now.get(Calendar.HOUR_OF_DAY) * 60 + now.get(Calendar.MINUTE)
        return if (from < to) minute >= from && minute < to else minute >= from || minute < to
    }

    fun vehicleMask(vehicle: String?): Int = when (vehicle?.lowercase(Locale.US)) {
        null -> 0
        "car" -> CAR
        "bike" -> BIKE
        "truck" -> TRUCK
        "bus" -> BUS
        else -> 16
    }
}
